galley/pkg/crd/validation/webhook.go

```diff
@@ -0,0 +1,354 @@
+
+import (
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/ghodss/yaml"
+	"github.com/howeyc/fsnotify"
+	admissionv1beta1 "k8s.io/api/admission/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+
+	"istio.io/istio/mixer/adapter"
+	"istio.io/istio/mixer/pkg/config"
+	mixerCrd "istio.io/istio/mixer/pkg/config/crd"
+	"istio.io/istio/mixer/pkg/config/store"
+	"istio.io/istio/mixer/pkg/lang/checker"
+	runtimeConfig "istio.io/istio/mixer/pkg/runtime/config"
+	"istio.io/istio/mixer/pkg/runtime/validator"
+	"istio.io/istio/mixer/pkg/template"
+	generatedTmplRepo "istio.io/istio/mixer/template"
+	"istio.io/istio/pilot/pkg/config/kube/crd"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pkg/log"
+)
+
+var (
+	runtimeScheme = runtime.NewScheme()
+	codecs        = serializer.NewCodecFactory(runtimeScheme)
+	deserializer  = codecs.UniversalDeserializer()
+)
+
+const (
+	watchDebounceDelay = 100 * time.Millisecond
+)
+
+// WebhookParameters contains the configuration for the Istio Pilot validation
+// admission controller.
+type WebhookParameters struct {
+	// MixerValidator implements the backend validator functions for mixer configuration.
+	MixerValidator store.BackendValidator
+
+	// PilotDescriptor provides a description of all pilot configuration resources.
+	PilotDescriptor model.ConfigDescriptor
+
+	// DomainSuffix is the DNS domain suffix for Pilot CRD resources,
+	// e.g. cluster.local.
+	DomainSuffix string
+
+	// Port where the webhook is served. Per k8s admission
+	// registration requirements this should be 443 unless there is
+	// only a single port for the service.
+	Port uint
+
+	// CertFile is the path to the x509 certificate for https.
+	CertFile string
+
+	// KeyFile is the path to the x509 private key matching `CertFile`.
+	KeyFile string
+
+	// HealthCheckInterval configures how frequently the health check
+	// file is updated. Value of zero disables the health check
+	// update.
+	HealthCheckInterval time.Duration
+
+	// HealthCheckFile specifies the path to the health check file
+	// that is periodically updated.
+	HealthCheckFile string
+}
+
+// Webhook implements the validating admission webhook for validating Istio configuration.
+type Webhook struct {
+	mu   sync.RWMutex
+	cert *tls.Certificate
+
+	// pilot
+	descriptor   model.ConfigDescriptor
+	domainSuffix string
+
+	// mixer
+	validator store.BackendValidator
+
+	healthCheckInterval time.Duration
+	healthCheckFile     string
+	server              *http.Server
+	watcher             *fsnotify.Watcher
+	certFile            string
+	keyFile             string
+}
+
+// CreateMixerValidator creates a mixer backend validator.
+func CreateMixerValidator(kubeconfig string) (store.BackendValidator, error) {
+	info := generatedTmplRepo.SupportedTmplInfo
```

Comment:
These seem problematic. Specifically, this depends on Mixer's own built-in templates, which effectively exposes Mixer internals to Galley. (which is also apparent from the imports in this file).

I wasn't expecting to pull this much Mixer code into Galley. Would it be better if we split this CR and do some refactoring before implementing the webhook in Galley?

Reply:
Per our offline discussion I'd prefer to reverse the order since config validation is P0 for Istio 1.0. I created https://github.com/istio/istio/issues/4887 to track the follow-up refactoring.

Reply:
@geeknoid 
FYI. Hopefully, after Mandar lands CRD normalization, we can start utilizing a declarative config check for adapter configs and reduce the scope of dependencies to Mixer to the pkg/ level.